core/src/main/scala/org/apache/spark/network/netty/NettyBlockTransferService.scala

```diff
@@ -48,6 +54,41 @@ class NettyBlockTransferService(conf: SparkConf, securityManager: SecurityManage
   private[this] var clientFactory: TransportClientFactory = _
   private[this] var appId: String = _
 
+  override def metricsCollection(executorId: String, executorMetrics: ExecutorMetrics): Unit = {
+    logDebug("starting a new thread for metricsCollection()")
+    new Thread("metric thread for executor " + executorId) {
+      override def run() {
+        while (true) {
+          getNettyMemMetrics(executorMetrics)
+          Thread.sleep(2000)
+        }
+      }
+    }.start()
+  }
+
+  private def getNettyMemMetrics(executorMetrics: ExecutorMetrics): Unit = {
+    val pooledAllocator = clientFactory.getPooledAllocator()
+    val directHeapSize: Long = sumOfMetrics(pooledAllocator.directArenas())
+    val onHeapSize: Long = sumOfMetrics(pooledAllocator.heapArenas())
+    executorMetrics.setNettyReadOnheapSize(onHeapSize)
+    executorMetrics.setNettyReadDirectheapSize(directHeapSize)
+    logInfo(s"current Netty directHeapSize is $directHeapSize, heapSize is $onHeapSize, " +
+        s"executer id is ${SparkEnv.get.blockManager.blockManagerId.executorId}")
+  }
+
+  private def sumOfMetrics(arenaMetricList: List[PoolArenaMetric]): Long = {
+    arenaMetricList.map { Arena =>
+      Arena.chunkLists().map { chunk =>
+        val itr: Iterator[PoolChunkMetric] = chunk.iterator()
+        var chunkSum: Long = 0
+        while (itr.hasNext) {
+          chunkSum += itr.next().chunkSize()
+        }
+        chunkSum
```

Comment:
I think you can also do `chunk.iterator.map{_.chunkSize()}.sum`.  You might need to `import scala.collections.JavaConverters._` and then do `chunk.iterator.asScala ...`, though I don't think its necessary.  `sum` on an iterator in scala will not create an intermediate collection (though it will do boxing).